### ISSUE TITLE
Java heap space error when writing > 2 GB file content as Blob into Oracle Table

### ISSUE THREAD

**Issue by harinivas-ganapathy** [JetBrains/Exposed#1617]
I have scenario to read a remote file of >2 GB file into an Oracle Blob Column. When ever the remote file size is > 1.2 GB I am am getting Java heap space error.

Below is my implementation? Can you please help?

```
object FileDownload: Table("FILE_DOWNLOAD") {
      private val fileID = integer("FILE_ID")
     val fileContent = blob("FILE_CONTENT")
}

val inStream = sftp.get(filePath)

transaction(db) {
    FileDownload.insert {
        it[fileContent] = ExposedBlob(inStream.readBytes())
    }
    inStream.close()
}
```

**Comment by AlexeySoshin** [comment]
Looking at the code, I think Exposed doesn't handle LOBs that well at the moment:
https://github.com/JetBrains/Exposed/blob/98bdca8a1563a34af9d3d7a357c9b0c0f5fae7e3/exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/statements/api/ExposedBlob.kt#L3

Here the Exposed Blob wrapper will create a copy of LOB in the memory.

Now, under the hood, when writing to the database it actually uses `inputStream`, which should read data in chunks, thus avoiding the OutOfMemory error you're getting:
https://github.com/JetBrains/Exposed/blob/16af55736dc9bec6f770faba4e49de98d880d0f9/exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/ColumnType.kt#L737

What you can try and do is to use raw statements and see if that solves your issue. 

```kotlin
transaction(db) {
    exec("insert into FILE_DOWNLOAD (FILE_CONTENT) values (?)",
            listOf(BlobColumnType() to inStream)
        )
}
```

If it does help, then I'd be happy to look into adding support of InputStreams for the `ExposedBlob` type.

**Comment by harinivas-ganapathy** [comment]
Thanks for the suggestion. I tried using raw statements, though I am not seeing any error, when I check the Oracle database I only see `null` value inserted into the Blob Column.

**Comment by AlexeySoshin** [comment]
Are you sure that your input stream is not-null?

**Comment by harinivas-ganapathy** [comment]
Yes I could confirm the remote file has 5GB content. Also I get the Java Heap error when trying to readBytes from the input stream.

**Comment by harinivas-ganapathy** [comment]
Also, when I tried for smaller sized file I only see null value in DB. But if I do `BlobColumnType() to inStream.readBytes()` for the same smaller file I could see the right value in the blob column.

**Comment by AlexeySoshin** [comment]
`readBytes()` will always attempt to read the entire file contents into the memory.

That's why you'll get OutOfMemory for larger files. 

But `BlobColumnType() to inStream` should work, although I could test it only on PostgreSQL

Two things to check:
1. `BlobColumnType() to inStream` with a small file. 
2.  What's the return type of `sftp.get(filePath)`? Is it `InputStream` or something more specific?

**Comment by Tapac** [comment]
It might become better with the next release. 
Now ExposedBlob stores inputStream by default, so please wait for the release and check will it helps or not.
Don't forget to replace `ExposedBlob(inStream.readBytes())` with `ExposedBlob(inStream)`

**Comment by Tapac** [comment]
@harinivas-ganapathy , please check with Exposed 0.41.1

**Comment by harinivas-ganapathy** [comment]
Also I noticed the `inStream.available()` (for local file) returns the same value 2147483647 as the amount of Bytes written into Blob Column

**Comment by oharaandrew314** [comment]
Is this issue resolved thanks to commit 367babcbf24f7bcaeb4b5e33c1b688cab38287cd?

**Comment by joc-a** [comment]
Hello @harinivas-ganapathy, could you please check if this issue is resolved with commit [367babc](https://github.com/JetBrains/Exposed/commit/367babcbf24f7bcaeb4b5e33c1b688cab38287cd)?
